DOCU-2423: Adds StatsD monitoring page under Kong in Production (#4260)
Kong/docs.konghq.com: src/gateway/kong-production/monitoring/statsd.md

---
title: Monitoring with StatsD
content_type: how-to
---

[StatsD](https://github.com/statsd/statsd) is a network daemon that collects
and aggregates performance metrics by listening on the network for 
text based statistics data, published by applications. 

This guide will help you setup a test {{site.base_gateway}} and
StatsD service. Then you will generate sample requests to {{site.base_gateway}} and
observe the collected monitoring data. 

### Prerequisites
* [Docker](https://docs.docker.com/get-docker/) is used to run StatsD and supporting services locally. 
* [curl](https://curl.se/) is used to send requests to {{site.base_gateway}}. `curl` is pre-installed on most systems.
* [Netcat](http://netcat.sourceforge.net/) is installed as `nc` on the system `PATH`. `nc` is used to send requests 
  to the StatsD management interface. `nc` is pre-installed on many systems.

### Configure StatsD monitoring

> 1. Install a test {{site.base_gateway}}:

Reviewer: ```suggestion
1. Install {{site.base_gateway}}:
```

---
title: Monitoring with StatsD
content_type: how-to
---

[StatsD](https://github.com/statsd/statsd) is a network daemon that collects
and aggreates performance metrics by listening on the network for simple 
text based statistics data published by applications. 

This guide will help you setup a test {{site.base_gateway}} and
StatsD service. Then you will generate sample requests to {{site.base_gateway}} and
observe the collected monitoring data.

### Prerequisites

> This guide assumes each of the following tools are installed locally. 

Reviewer 2: ```suggestion
This guide assumes the following tools are installed locally:
```

> * [Docker](https://docs.docker.com/get-docker/) is used to run Kong, the supporting database, and StatsD locally. 

Reviewer 2: ```suggestion
* [Docker](https://docs.docker.com/get-docker/) is used to run {{site.base_gateway}}, the supporting database, and StatsD locally. 
```

> * [curl](https://curl.se/) is used to send requests to the gateway. Most systems come with `curl` pre-installed.

Reviewer 2: ```suggestion
* [curl](https://curl.se/) is used to send requests to {{site.base_gateway}}. `curl` is pre-installed on most systems.
```

> * [Netcat](http://netcat.sourceforge.net/) installed as `nc` on the `PATH`. `nc` is used to send requests 
>   to the StatsD management interface. Many systems come with `nc` pre-installed.

Reviewer 2: ```suggestion
* [Netcat](http://netcat.sourceforge.net/) installed as `nc` on the `PATH`. `nc` is used to send requests 
  to the StatsD management interface. `nc` is pre-installed on many systems.
```

> ### Guide

Reviewer 2: ```suggestion
### Configure StatsD monitoring
```

> 1. Run the following command to start {{site.base_gateway}} using Docker:

Reviewer 2: @rspurgeon What happens if you already have Gateway installed and you try to run this script? I was kind of thinking that the audience of this doc might already have Gateway installed. If so, would it be better to tell them which routes/services/etc. to configure so they can use this guide? Or we could provide the option like:

1. Do one of the following:
    - If you don't have Kong installed, run this quickstart script:
    - If you've already installed Kong, make sure you add these services/routes that we'll be using for this guide:

Author: The quickstart script is designed to rebuild the test gateway if it already exists.

IMO we should just give the user the necessary instructions inline of page.  It's a single line of code to get the test gateway, so i think it's a distraction to redirect them to another page.  The script also sets up the mock service which aids in the later instructions, otherwise we have to write confusing "if you have this do this if you don't have this do that" kind of copy.

Reviewer 3: I think Diana's idea is a good one. There's nothing wrong with having the script as a prerequisite, or providing the alternative to add a list of routes and services that they will be using in this guide. 

While the quickstart script and the entities it creates are great resources, sticking this in as an absolute requirement gates the guide for anyone with an existing installation. 

> otherwise we have to write confusing "if you have this do this if you don't have this do that" kind of copy.

We can write this once, in the prereq, and then the rest of the guide can just assume they have the routes/services.

> The quickstart script is designed to rebuild the test gateway if it already exists.

That seems like a bad idea, if the person isn't expecting it. If we don't give them the option to add this to their existing installation, then we're losing potential users.

As much as I like the idea of "every page is page one", starting with an installation of kong gateway for a guide like this is going too far. We have to remember that our audience is varied, and users of all stages in their Kong journey could be wanting to do something like set up statsd - by adding a step to install gateway anew, we're essentially saying "you can't actually do this in your installation, you have to start over".

Author: In my experience this kind of soup-to-nuts quickstart documentation is best for getting up and running with concepts, otherwise the documentation becomes littered with conditionals for the reader (click this link if you have this or do this if you don't).  Documentation pages that are simple, complete, and to the point is the best in my personal opinion.

The task at hand is to learn about the StatsD plugin and the concepts around it.  As a potential reader of this documentation, I would assume I could apply the techniques to my own installation, or else why would it exist at all? 

This is why people use Docker for development, ephemeral things to test that are easily disposed and recreated. The script gives them a quick test environment to work with, named in a particular way that can be disposed of with no cost.  They should have no expectation about it that would surprise them.  

When testing new concepts, it's less common to want to add to an existing service, one that is likely already serving another purpose. You spin up something new to learn the objective, take the education and then apply it. I wouldn't expect lose any potential users over this workflow.  

All that said, I think some dialog at the beginning of the document that says, "If you don't have a Kong Gateway, run this script" and keep the script instruction _in line of the page_ would be fine.

Then we also need to say something like... "If you're using an existing gateway, modify the connection information in the following command."  Otherwise we need to use a variable or something which is currently represented in the doc as `localhost:8000-1`.   As Diana also mentioned, we'd also need to add something to inform the user to add a new route and service to match the document, or tell them to modify the commands as needed.

In my personal opinion, it adds a lot of extra cognitive overhead for the reader who is very likely learning this on a test gateway they will create. Regardless, please feel free to make the desired changes to the document as the team desires to get it complete for the release.

   ```sh
   curl -Ls get.konghq.com/quickstart | sh -s
   ```

   You should see the following message when {{site.base_gateway}} is ready:

   ```text
   ✔ Kong is ready!
   ```

>    The script has also installed a mock service to make testing easier. You will use this 

Reviewer 2: ```suggestion
   The script also installs a mock service to make testing easier. You will use this 
```

   service later to generate metrics data. You can send a sample request with:

   ```sh
   curl localhost:8000/mock/requests
   ```

1. Run a StatsD container to capture monitoring data:

   ```sh
   docker run -d --rm -p 8126:8126 \
     --name kong-quickstart-statsd --network=kong-quickstart-net \
     statsd/statsd:latest
   ```

1. Install the [StatsD {{site.base_gateway}} plugin](/hub/kong-inc/statsd/), 
   configuring the hostname and port of the listening StatsD service:

   ```sh
   curl -X POST http://localhost:8001/plugins/ \
       --data "name=statsd"  \
       --data "config.host=kong-quickstart-statsd" \
       --data "config.port=8125"
   ```

   You should receive a JSON response with the details of the installed plugin.

> 1. Generate sample traffic to the mock service. This will allow you to observe 
>    metrics generated from the StatsD plugin. The following command will generate 60 
>    requests over 1 minute. Run the following in a new terminal:

Reviewer 2: ```suggestion
1. Generate sample traffic to the mock service. This allows you to observe 
   metrics generated from the StatsD plugin. The following command generates 60 
   requests over one minute. Run the following in a new terminal:
```

   ```sh
   for _ in {1..60}; do {curl localhost:8000/mock/request; sleep 1; } done
   ```

1. Query the StatsD management interface to see the collected metrics from {{site.base_gateway}}:

   ```sh
   echo "counters" | nc localhost 8126
   ```

   You should see a response similar to the following:

   ```text
   {
     'statsd.bad_lines_seen': 0,
     'statsd.packets_received': 56,
     'statsd.metrics_received': 56,
     'kong.mock.request.count': 7,
     'kong.mock.request.status.200': 7,
     'kong.mock.request.status.total': 7
   }
   END
   ```

> The [Kong Hub StatsD Plugin](/hub/kong-inc/statsd/) 
> contains the full documentation on the plugin usage and configuration

Reviewer 2: ```suggestion
See the [StatsD plugin](/hub/kong-inc/statsd/) 
documentation for more information about how to use and configure the plugin.
```

> ### Cleanup

Reviewer 2: ```suggestion
### Remove the test StatsD service 
```

Author: The cleanup does more than just remove the statsd service, it cleans up the gateway, database, and docker network as well

> ### Remove the test StatsD service 

Author: I still think "cleanup" here is the best word to encompass what the steps do.

I think we might be debating the difference between a how-to guide and a reference document.  I might be mis-informed about the goal of this page, but I don't view this as the reference material for StatsD. I view this as a conceptual How-to with the goal to get the user to understand that Monitoring with StatsD is possible and what are the various concepts and components involved.

We have Docker as a prereq here, and we instruct the user to run StatsD in Docker. If they are going to connect that to a "Production" gateway (or any Gateway not in the same Docker netwrok), there are other factors we might have to document.  Network connectivity for example might not work "OOTB" from StatsD to GW when it's running in a container.  I don't think in the area of conceptual learning, people are / should be connecting to production services and I'm not sure we should solve for that case here.

> Once you are done experimenting with StatsD and {{site.base_gateway}}, use the following
> commands to stop and remove the services created in this guide:

Reviewer 2: ```suggestion
Once you are done experimenting with StatsD and {{site.base_gateway}}, you can use the following
commands to stop and remove the services created in this guide:
```

```sh
docker stop kong-quickstart-statsd
curl -Ls get.konghq.com/quickstart | sh -s -- -d
```

> ### More Information

Reviewer 2: ```suggestion
### More information
```

> * [How to monitor with Prometheus](/gateway/{{page.gateway_version}}/kong-production/monitoring/prometheus/) 
> provides a guide to using [Prometheus](https://prometheus.io/docs/introduction/overview/) for monitoring with the 
> [{{site.base_gateway}} Plugin](/hub/kong-inc/prometheus/)

Reviewer 2: ```suggestion
* [How to monitor with Prometheus](/gateway/{{page.kong_version}}/kong-production/monitoring/prometheus/) 
describes how to use [Prometheus](https://prometheus.io/docs/introduction/overview/) to monitor {{site.base_gateway}} using the  
[Prometheus plugin](/hub/kong-inc/prometheus/).
```

> * See the [Tracing API Reference](/gateway/{{page.kong_version}}/kong-production/tracing/api/) for information
> on {{site.base_gateway}}'s tracing capabilites

Reviewer 2: ```suggestion
* See the [Tracing API Reference](/gateway/{{page.kong_version}}/kong-production/tracing/api/) for information
about {{site.base_gateway}}'s tracing capabilities.
```

---
title: Monitoring with StatsD
content_type: how-to
---

[StatsD](https://github.com/statsd/statsd) is a network daemon that collects
and aggreates performance metrics by listening on the network for simple 
text based statistics data published by applications. 

This guide will help you setup a test {{site.base_gateway}} and
StatsD service. Then you will generate sample requests to {{site.base_gateway}} and
observe the collected monitoring data. 

### Prerequisites
* A test {{site.base_gateway}} instance is installed using the following script:
   ```sh
   curl -Ls get.konghq.com/quickstart | sh -s -- -m
   ```
   The `-m` flag instructs the script to install a mock service that is used in this guide to generate sample metrics.
   {:.note}

>    > **Note:** This guide's instructions use a test {{site.base_gateway}} instance to demonstrate how the StatsD plugin can be used to collect metrics. If you want to use an existing {{site.base_gateway}} instance for this guide, you must modify the connection information in the commands.

Reviewer 2: @rspurgeon @lena-larionova I'm not sure why GitHub removed our conversation, but I added this and was wondering what you think. Should we provide additional information in the note about what services/routes/etc. they should add to an existing instance, or would users know enough to know what to modify in the commands?

Author: I'm in favor of this.  I do think if someone is going to be connecting StatsD, from inside Docker, to an existing Gateway, they likely need to have enough knowledge to do that without a lot of very specific guidance from them.  Their gateway could be anywhere, so we can't give them exact guidance.

The only other change I might make is, instead of it being a note after the user runs the script, but this guidance before the script and tell them it's optional and why.  So i'd move this up to the instructions for Step 1

Author: I didn't notice that this step got moved to Prerequisites.  IMO this isn't a prereq, it's a step in the guide.  We may choose to make that step optional, but I think prereqs are tools that must be installed or other system setups that are required to proceed.    If this were a pre-req, it would need to come after Docker or curl becuase it uses both of those tools.  I still very much prefer that this step be part of the guide with text to make it optional if desired.

---
title: Monitoring with StatsD
content_type: how-to
---

[StatsD](https://github.com/statsd/statsd) is a network daemon that collects
and aggreates performance metrics by listening on the network for simple 
text based statistics data published by applications. 

This guide will help you setup a test {{site.base_gateway}} and
StatsD service. Then you will generate sample requests to {{site.base_gateway}} and
observe the collected monitoring data. 

### Prerequisites
* [Docker](https://docs.docker.com/get-docker/) is installed locally. It is used to run {{site.base_gateway}}, the supporting database, and StatsD locally. 

>    If you will be using an existing {{site.base_gateway}} instance instead of the test instance, you may need configure Docker to ensure features like network connectivity function correctly.  

Reviewer 2: ```suggestion
   If you will be using an existing {{site.base_gateway}} instance instead of the test instance, you may need configure Docker to ensure Docker and {{site.base_gateway}} are running on the same network.  
```
@rspurgeon For your comment about UDP, would something like this work?

Author: @cloudjumpercat I've just pushed a revision, please take a look.  I think since we're making the Docker based GW optional, we have to be vague about the instructions for the user wishing to use their own.  UDP networking rules are not easily described here, but it won't be accurate to say that the StatsD in Docker and the existing GW have to be on the "same network" because StatsD is ran in a new Docker network.

In fact, I just realized, the StatsD docker run command is not going to work OOTB if the user does not use the script because it uses a new network which won't exist (--network=kong-quickstart-net), so the user will need to navigate this with our intentionally vague instructions here.

* [curl](https://curl.se/) is installed locally. It is used to send requests to {{site.base_gateway}}. `curl` is pre-installed on most systems.
* [Netcat](http://netcat.sourceforge.net/) is installed locally as `nc` on the `PATH`. `nc` is used to send requests 
  to the StatsD management interface. `nc` is pre-installed on many systems.

### Configure StatsD monitoring

{:.note}
   > **Note:** This guide's instructions use a test {{site.base_gateway}} instance to demonstrate how the StatsD plugin can be used to collect metrics. If you want to use an existing {{site.base_gateway}} instance for this guide, you must modify the connection information in the commands.

> 1. Optional: Install {{site.base_gateway}} using the following script:

Reviewer 2: @rspurgeon I moved the note and script back to the steps. Does this look good to you?

Author: @cloudjumpercat  I've just pushed a revision, please take a look

   ```sh
   curl -X POST http://localhost:8001/plugins/ \
       --data "name=statsd"  \
       --data "config.host=kong-quickstart-statsd" \
       --data "config.port=8125"
   ```

   You should receive a JSON response with the details of the installed plugin.

1. Generate sample traffic to the mock service. This allows you to observe 
   metrics generated from the StatsD plugin. The following command generates 60 
   requests over one minute. Run the following in a new terminal:

   ```sh
   for _ in {1..60}; do {curl localhost:8000/mock/request; sleep 1; } done
   ```

1. Query the StatsD management interface to see the collected metrics from {{site.base_gateway}}:

   ```sh
   echo "counters" | nc localhost 8126
   ```

   You should see a response similar to the following:

   ```text
   {
     'statsd.bad_lines_seen': 0,
     'statsd.packets_received': 56,
     'statsd.metrics_received': 56,
     'kong.mock.request.count': 7,
     'kong.mock.request.status.200': 7,
     'kong.mock.request.status.total': 7
   }
   END
   ```

See the [StatsD plugin](/hub/kong-inc/statsd/) 
documentation for more information about how to use and configure the plugin.

> ### Clean up the test instance 

Reviewer 2: @rspurgeon Would this heading work for you? I wanted to add more to "clean up" just so users can know what the section will cover at a glance.

Author: @cloudjumpercat I pushed a revision, please take a look.  I personally think just "Clean up" is sufficient, but if you want additional text "test instance" isn't exactly accurate, because the instructions inform the user to cleanup the test kong gateway, database and network along with the StatsD service.  I modified to be vague and say "running software"
